Q21. In a jupyter notebook solve the following question. Please upload the notebook to GitHub and provide the link submission box below.

 

__int()__: Constructor that takes as input a pair of Point objects that represent the ends points of the line segment

Length():: returns the length if the segment 

Slope() returns the slope of the segment of none if the slope is unbounded 

 

>>> p1 = Point(3,4)

>>> p2 = Point()

>>> s = Segment(p1,p2)

>>> s.length()

5.0

>>> s.slope()

0.75

In [4]:
class Point:
    def __init__(self,x,y):
        self.x = x
        self.y = y

class Segment:
    def __init__(self, p1, p2):
        self.p1 = p1
        self.p2 = p2

    def length(self):
        # using the distance formula
        distance = ((self.p1.x - self.p2.x) ** 2 + (self.p1.y - self.p2.y) ** 2) ** 0.5
        return distance

    def slope(self):
        if self.p1.x == self.p2.x: # check if it is a vqertical line
            return None
        slope = (self.p2.y - self.p1.y) / (self.p2.x - self.p1.x) # using slope formula
        return slope

# Example usage
p1 = Point(3, 4)
p2 = Point()
s = Segment(p1, p2)

print("Length:", s.length())
print("Slope:", s.slope())


Length: 5.0
Slope: 1.3333333333333333


### as shown in the Question the slope is 0.75.  But when we apply slope formula to the given input, we get 1.333

Q22. In a jupyter notebook solve the following question using both python and SQL. Please upload the notebook to GitHub and provide the link submission box below.



+---------------+---------+
| Column Name   | Type    |
+---------------+---------+
| id            | int     |
| recordDate    | date    |
| temperature   | int     |
+---------------+---------+
id is the column with unique values for this table.
This table contains information about the temperature on a certain day.
 
 

Write a solution to find all dates' Id with higher temperatures compared to its previous dates (yesterday).

Return the result table in any order.

Please use the following input Weather table for your solution. To receive full marks you will need to create the database, create the table, insert the data below and execute the SQL query.

 

Input: 
Weather table:
+----+------------+-------------+
| id | recordDate | temperature |
+----+------------+-------------+
| 1  | 2015-01-01 | 10          |
| 2  | 2015-01-02 | 25          |
| 3  | 2015-01-03 | 20          |
| 4  | 2015-01-04 | 30          |
+----+------------+-------------+
Output: 
+----+
| id |
+----+
| 2  |
| 4  |
+----+
Explanation: 
In 2015-01-02, the temperature was higher than the previous day (10 -> 25).
In 2015-01-04, the temperature was higher than the previous day (20 -> 30).

In [6]:
!pip install ipython-sql


     |████████████████████████████████| 41 kB 467 kB/s eta 0:00:01
     |████████████████████████████████| 2.1 MB 1.4 MB/s eta 0:00:01
  Attempting uninstall: sqlalchemy
    Found existing installation: SQLAlchemy 1.4.32
    Uninstalling SQLAlchemy-1.4.32:
      Successfully uninstalled SQLAlchemy-1.4.32


In [10]:
%load_ext sql


The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [16]:
%%sql

CREATE TABLE Weather
(
    id int,
    recordDate DATE,
    temperature int
);

INSERT INTO Weather (id, recordDate, temperature) VALUES
    (1, '2015-01-01', 10),
    (2, '2015-01-02', 25),
    (3, '2015-01-03', 20),
    (4, '2015-01-04', 30);


 * sqlite:///your_database.db
Done.
4 rows affected.


[]

In [40]:
%%sql

WITH WeatherWithPrev AS (
    SELECT w.id AS current_id, w.recordDate AS current_date, w.temperature AS current_temp,LAG(w.temperature) 
    OVER (ORDER BY w.recordDate) AS prev_temp
    FROM Weather w
)

SELECT current_id AS id FROM WeatherWithPrev WHERE current_temp > prev_temp;


 * sqlite:///your_database.db
Done.


[(2,), (4,)]

Q23. In a jupyter notebook solve the following question using python. Please upload the notebook to GitHub and provide the link submission box below.



Develop a recursive function tough() that takes two nonnegative integer arguments and outputs a pattern as shown below. Hint: The first argument represents the indentation of the pattern, where the second argument -- always a pattern of 2 indicates the number *s in the longest line of *s in the pattern


In [39]:
def f(indent, num):
    if(num > 0): #If num is 0 or less, the recursion stops
        f(indent, num//2); #recurssive call
        print(" "*indent + "*"*num); #printing pattern
        f(indent + num//2, num//2);  #second recurssive call

In [36]:
f(0,0)


*
**
 *
****
  *
  **
   *


In [37]:
f(0,2)

*
**
 *


In [38]:
f(0,4)

*
**
 *
****
  *
  **
   *
